Doc/howto/isolating-extensions.rst

```diff
@@ -438,9 +438,20 @@ GC-tracked objects need to be allocated using GC-aware functions.
 
 If you use use :c:func:`PyObject_New` or :c:func:`PyObject_NewVar`:
 
-- Get and call type's :c:member:`~PyTypeObject.tp_alloc` slot , if possible.
-- If not possible (e.g. inside a custom ``tp_alloc``),
-  call :c:func:`PyObject_GC_New` or :c:func:`PyObject_GC_NewVar`.
+- Get and call type's :c:member:`~PyTypeObject.tp_alloc` slot, if possible.
+  That is, replace ``TYPE *o = PyObject_New(TYPE, typeobj)`` by::
+
+      TYPE *o = typeobj->tp_alloc(typeobj, 0);
```

Comment:
It's a shame `PyObject_New()` (and `PyObject_NewVar()`) don't take care of this.  Assuming they can't, would it make sense to provide a new function that does the right thing?  That way we avoid people relying on `tp_alloc`.

<details>
<summary>For example:</summary>

```C
// Include/object.c
PyAPI_FUNC(PyObject *) _PyObject_NewEx(PyTypeObject *);
PyAPI_FUNC(PyObject *) _PyObject_NewVarEx(PyTypeObject *, Py_ssize_t);

#define PyObject_NewEx(type, typeobj) \
    ((type *)_PyObject_NewEx(typeobj))
#define PyObject_NewVarEx(type, typeobj, n) \
    ((type *)_PyObject_NewVarEx(typeobj, n))
```

```C
// Objects/object.c
PyObject *
_PyObject_NewEx(PyTypeObject *tp)
{
    return tp->tp_alloc(tp, 0);
}

PyObject *
_PyObject_NewVarEx(PyTypeObject *tp, Py_ssize_t nitems)
{
    return tp->tp_alloc(tp, nitems);
}
```

or even:

```C
// Objects/object.c
PyObject *
_PyObject_NewEx(PyTypeObject *tp)
{
    if (_PyType_IS_GC(tp) && tp->tp_alloc == PyBaseObject.tp_alloc) {
        return _PyObject_GC_New(tp);
    }
    return tp->tp_alloc(tp, 0);
}

PyObject *
_PyObject_NewVarEx(PyTypeObject *tp, Py_ssize_t nitems)
{
    if (_PyType_IS_GC(tp) && tp->tp_alloc == PyBaseObject.tp_alloc) {
        return _PyObject_GC_NewVar(tp, nitems);
    }
    return tp->tp_alloc(tp, nitems);
}
```

</details>

Reply:
There are many things that could be improved with new API. I'd like to keep this PR focused on the HOWTO, which will need to cover old versions anyway.

More generally, I think it'd be good to provide “caller” API for all the slots. Something like that will be necessary if we ever want to start changing the slot signatures.